typed_sql/doc/03_writing_queries.md

```diff
@@ -0,0 +1,590 @@
+    .orderBy((b) => [(b.stock, Order.descending)])
+    .select((b) => (b.title, b.stock))
+    .fetch();
+
+check(result).deepEquals([
+  // title, stock
+  ('Vegetarian Dining', 42),
+  ('Hiding Eggs for dummies', 12),
+  ('Are Bunnies Unhealthy?', 10),
+  ('Vegan Dining', 3),
+  ('Cooking with Chocolate Eggs', 0),
+]);
+```
+
+The equivalent SQL query would look something like:
+```sql
+SELECT
+  title,
+  stock
+FROM books
+ORDER BY stock DESC
+LIMIT 3;
+```
+
+You must always provide `Order.ascending` or `Order.descending` for entries in
+the return value from the callback in `.orderBy`. You can sort by any
+`Expr<Comparable>`, typically `Expr<int>`, `Expr<double>`, `Expr<DateTime>`,
+`Expr<String>`, or, `Expr<bool>`. Again, you can sort by expressions other than
+just fields, including subqueries, but you might want to consider the
+performance implications before using a subquery for ordering.
+
+
+## Limit and offset with `.limit` and `.offset`
+We can use `.limit` and `.offset` to return a subset of rows from a query.
+The following example builds on top of the previous example, sorting by
+stock in descending order, skipping the first two rows and returning at-most 3
+rows.
+
+```dart bookstore_test.dart#query-orderby-offset
+final result = await db.books
+    .orderBy((b) => [(b.stock, Order.descending)])
+    .select((b) => (b.title, b.stock))
+    // The order in which .orderBy, .offset, .limit appears is significant.
+    .offset(2)
+    .limit(3)
+    .fetch();
+
+check(result).deepEquals([
+  // title, stock
+  ('Are Bunnies Unhealthy?', 10),
+  ('Vegan Dining', 3),
+  ('Cooking with Chocolate Eggs', 0),
+]);
+```
+
+The equivalent SQL query would look something like:
+```sql
+SELECT
+  title,
+  stock
+FROM books
+ORDER BY stock DESC
+OFFSET 2
+LIMIT 3;
+```
+
+> [!WARNING]
+> The **ordering** of `.orderBy`, `.offset` and `.limit` is significant.
+> Meaning that `.limit(3).offset(3)` will _always_ result in an empty query!
+
+You may apply `.limit` and `.offset` more than once, but they do stack on-top of
+each other. So `.limit(3).limit(4)` will at-most return 3 rows, and
+`.offset(2).offset(1)` will skip the first 3 rows.
+
+
+## Point queries with `.byKey`, `.first` and `db.select()`
+If we want to lookup a row by _primary key_ `package:typed_sql` will generate
+a convenient `.byKey()` method. If we wanted to look up the row in books with
+`bookId = 1` we could simple write:
+
+```dart bookstore_test.dart#books-bykey
+final book = await db.books.byKey(bookId: 1).fetch();
+if (book == null) {
+  throw Exception('Book not found');
+}
+
+check(book.title).equals('Are Bunnies Unhealthy?');
+```
+
+The astute reader might realize that unlike previous query examples `.fetch()`
+does not return a `Future<List<Book>>`. This is because `.byKey` returns a
+`QuerySingle<(Expr<Book>,)>`, meaning that the `.fetch()` extension method knows
+that the query returns _at-most one row_. Thus, `.fetch()` can simple return
+`Future<Book?>`.
```

Comment:
```suggestion
Unlike previous query examples `.fetch()`
does not return a `Future<List<Book>>`. This is because `.byKey` returns a
`QuerySingle<(Expr<Book>,)>`, a query type that is guaranteed to return _at-most one_ row. Thus, `.fetch()` simply returns `Future<Book?>`.
```

Reply:
I don't like the word "guaranteed", it's "defined as".